core/services/ocr2/plugins/ocr2keeper/evm21/upkeepstate/store.go

```diff
@@ -200,13 +214,34 @@ func (u *upkeepStateStore) upsertStateRecord(ctx context.Context, workID string,
 
 	u.cache[workID] = record
 
-	return u.orm.InsertUpkeepState(persistedStateRecord{
+	u.pendingRecords = append(u.pendingRecords, persistedStateRecord{
 		UpkeepID:            utils.NewBig(upkeepID),
 		WorkID:              record.workID,
 		CompletionState:     uint8(record.state),
 		IneligibilityReason: reason,
 		InsertedAt:          record.addedAt,
-	}, pg.WithParentCtx(ctx))
+	})
+
+	if len(u.pendingRecords) >= flushSize {
+		u.flush(ctx)
+	}
+
+	return nil
+}
+
+func (u *upkeepStateStore) flush(ctx context.Context) {
+	u.wg.Add(1)
+
+	cloneRecords := make([]persistedStateRecord, len(u.pendingRecords))
+	copy(cloneRecords, u.pendingRecords)
+	u.pendingRecords = []persistedStateRecord{}
```

Comment:
would we need a lock on `u.pendingRecords` here? since it can be written to in `upsertStateRecord`